.github/workflows/build.yaml

```diff
@@ -55,9 +56,54 @@ jobs:
         run: |
           make test
         if: env.GIT_DIFF
+
+  set-environment:
+    runs-on: ubuntu-latest
+    needs: [golangci, test-unit]
+    outputs:
+      env-variable: ${{ steps.set-env-var.outputs.patch_env }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0  # Note: This fetches all branches and tags
+
+      - name: Check base ref
+        run: |
+          BASE_REF=$(git describe --contains --all HEAD)
+          echo "BASE_REF=$BASE_REF" >> $GITHUB_ENV
+
+      - name: Set ENV variable
+        id: set-env-var
+        run: |
+          PATCH_ENV="unknown" # Default value
+          case $GITHUB_REF in
+            refs/tags/*)
+              TAG_COMMIT=$(git rev-list -n 1 ${{ github.ref }})
+              BRANCHE=$(git branch -r --contains $TAG_COMMIT | sed 's/ *origin\///' | grep -v "HEAD" | head -n 1)
+              if [ "$BRANCHE" == "main" ]; then
+                PATCH_ENV="production"
+              fi
+              ;;
+            refs/heads/main)
+              PATCH_ENV="non-prod"
+              ;;
+          esac
+          echo "PATCH_ENV=$PATCH_ENV" >> $GITHUB_ENV
+          echo "BRANCHE=$BRANCHE" >> $GITHUB_ENV
+          echo "::set-output name=patch_env::$PATCH_ENV" 
+
+      - name: Debug
+        run: |
+          echo "Env: ${PATCH_ENV}"
+          echo "Ref: ${{ github.ref }}"
+          echo "GitHub Ref: $GITHUB_REF"
+          echo "Base Ref: $BASE_REF"
+          echo "Branch: $BRANCHE"
+
   build:
-    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
-    needs: test-unit
+    if: (github.event_name == 'push') && (needs.set-environment.outputs.env-variable == 'non-prod')
+    needs: [golangci, test-unit, set-environment]
```

Comment:
`test-unit` has `golangci` as dependency already, do not need to repeat it here, only `golangci` should be ok.